### PR TITLE
CB-16558 - Reduce the size of structuredevent and cdp_structured_event tables

### PR DESCRIPTION
**Pull request by gregito**

### CODE REVIEW

File: cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java [hunk 1]
```diff
@@ -28,25 +21,41 @@
 import com.sequenceiq.cloudbreak.common.type.ComponentType;
 import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
 import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponent;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponentHistory;
 import com.sequenceiq.cloudbreak.domain.view.ClusterComponentView;
+import com.sequenceiq.cloudbreak.repository.ClusterComponentHistoryRepository;
 import com.sequenceiq.cloudbreak.repository.ClusterComponentRepository;
 import com.sequenceiq.cloudbreak.repository.ClusterComponentViewRepository;
+import org.apache.commons.codec.binary.Base64;
```
Comment (bergerdenes) [inline cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java:29]: Please set a correct formatter, see for example line 35 should not be after these imports.
Please use the one included with the project itself

File: cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/common/TestUtil.java [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+    }
+
+    public static Network network() {
+        return network("10.0.0.1/16");
+    }
+
+    public static Network network(String subnet) {
+        Network network = new Network();
+        network.setSubnetCIDR(subnet);
+        //        network.setAddressPrefixCIDR(DUMMY_ADDRESS_PREFIX_CIDR);
```
Comment (bergerdenes) [inline cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/common/TestUtil.java:239]: dead code

File: cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/service/clustercomponent/ClusterComponentConfigProviderTestBase.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    private ClusterComponentHistoryRepository mockClusterComponentHistoryRepository;
+
+    @Mock
+    private AuditReader mockAuditReader;
+
+    private ClusterComponentConfigProvider underTest;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
Comment (bergerdenes) [inline cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/service/clustercomponent/ClusterComponentConfigProviderTestBase.java:46]: Please use @ExtendWith on class if possible

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpService.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    public void cleanUpStructuredEvents() {
+        LOGGER.debug("About to start fetch and clean up structured events.");
+        boolean success = true;
+        try {
+            Set<Long> idsForChecking = new HashSet<>();
+            idsForChecking.addAll(clusterService.findDeletedClusters().stream().map(Cluster::getId).collect(toSet()));
+            idsForChecking.addAll(stackService.findDeletedStackIds());
+            LOGGER.debug("About to request the deletion of {} {} that belogs to the following resource IDs: {}",
+                    StructuredEventEntity.class.getSimpleName(), idsForChecking.size() > 1 ? "entries" : "entry",
+                    String.join(",", idsForChecking.stream().map(Object::toString).collect(toSet())));
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpService.java:65]: I think (but you an experiment as well) that if you have a Set type variable and you call toString() on it, that will produce the same (or similar) result, good enough for logging.

File: core/src/test/java/com/sequenceiq/cloudbreak/service/cluster/flow/ClusterTerminationServiceTest.java [hunk 5]
```diff
@@ -65,20 +73,39 @@ class ClusterTerminationServiceTest {
...
+        prepareMockForFinalization();
+
+        underTest.finalizeClusterTermination(STACK_ID, FORCE);
+
+        assertThat(cluster.getStack()).isNotNull();
+        assertThat(cluster.getStack()).isEqualTo(cluster.getStack());
+    }
+
+    @Test
+    void testCleanUpClusterComponentsWhen() throws TransactionExecutionException {
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/service/cluster/flow/ClusterTerminationServiceTest.java:93]: when? ...

**Review by bergerdenes — CHANGES_REQUESTED** [review]
some small items

File: cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java [hunk 6]
```diff
@@ -199,11 +208,26 @@ public void restorePreviousVersion(ClusterComponent clusterComponent) {
         }
     }
 
+    public void cleanUpDetachedEntries() {
+        LOGGER.debug("About to fetch detached {} entries", ClusterComponentHistory.class.getSimpleName());
+        Set<ClusterComponentHistory> detachedEntries = clusterComponentHistoryRepository.findDetachedEntries();
```
Comment (topolyai5) [inline cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java:213]: unnecessary to fetch the all of the cluster component to check the count. Could you please refactor this?

**Review by topolyai5 — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java [hunk 7]
```diff
@@ -85,15 +89,15 @@ public boolean isEnabled() {
 
     @Override
     public <T extends StructuredEvent> List<T> getEventsForWorkspaceWithType(Workspace workspace, Class<T> eventClass) {
-        List<StructuredEventEntity> events = legacyStructuredEventRepository.findByWorkspaceAndEventType(workspace, StructuredEventType.getByClass(eventClass));
+        List<StructuredEventEntity> events = repository.findByWorkspaceAndEventType(workspace, StructuredEventType.getByClass(eventClass));
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java:92]: why do you rename it?

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java [hunk 8]
```diff
@@ -35,17 +30,26 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Component;
 
 @Component
 public class LegacyStructuredEventDBService extends AbstractWorkspaceAwareResourceService<StructuredEventEntity> implements LegacyStructuredEventService {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(LegacyStructuredEventDBService.class);
 
+    private static final int THREE_MONTHS = 86400 * 30 * 3;
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java:44]: Could you please try this with Duration(3, MONTH) or stuff like this

File: cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java [hunk 9]
```diff
@@ -199,11 +208,26 @@ public void restorePreviousVersion(ClusterComponent clusterComponent) {
         }
     }
 
+    public void cleanUpDetachedEntries() {
+        LOGGER.debug("About to fetch detached {} entries", ClusterComponentHistory.class.getSimpleName());
+        long detachedEntries = clusterComponentHistoryRepository.findDetachedEntries();
+        if (detachedEntries > 0) {
+            LOGGER.debug("{} detached {} entry has found that is about to get clean up", detachedEntries, ClusterComponentHistory.class.getSimpleName());
+            long deleted = clusterComponentHistoryRepository.deleteByClusterIsNullOrCluster_StatusIs(DELETE_COMPLETED);
```
Comment (topolyai5) [inline cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java:216]: we don't use the underscore, could you please replace it with `And`?

File: cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java [hunk 10]
```diff
@@ -29,25 +22,42 @@
 import com.sequenceiq.cloudbreak.common.type.ComponentType;
 import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
 import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponent;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponentHistory;
 import com.sequenceiq.cloudbreak.domain.view.ClusterComponentView;
+import com.sequenceiq.cloudbreak.repository.ClusterComponentHistoryRepository;
 import com.sequenceiq.cloudbreak.repository.ClusterComponentRepository;
 import com.sequenceiq.cloudbreak.repository.ClusterComponentViewRepository;
 import com.sequenceiq.cloudbreak.util.VersionNormalizer;
+import org.apache.commons.codec.binary.Base64;
```
Comment (bergerdenes) [inline cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/service/ClusterComponentConfigProvider.java:31]: the import order is _still_ incorrect. please update your format for jJava as attached in the project under `config/idea/idea_settings.jar`

File: cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/service/clustercomponent/ClusterComponentConfigProviderTestBase.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    private ClusterComponentHistoryRepository mockClusterComponentHistoryRepository;
+
+    @Mock
+    private AuditReader mockAuditReader;
+
+    private ClusterComponentConfigProvider underTest;
+
+    @BeforeEach
+    void setUp() {
+        underTest = new ClusterComponentConfigProvider(mockComponentRepository, mockComponentViewRepository, mockClusterComponentHistoryRepository,
```
Comment (bergerdenes) [inline cluster-api/src/test/java/com/sequenceiq/cloudbreak/cluster/service/clustercomponent/ClusterComponentConfigProviderTestBase.java:43]: `@InjectMocks` ?

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java [hunk 12]
```diff
@@ -184,4 +199,13 @@ private Stack getStackByCrn(Long workspaceId, String crn) {
         return Optional.ofNullable(stackService.getByCrnInWorkspace(crn, workspaceId))
                 .orElseThrow(notFound(Crn.safeFromString(crn).getResourceType().toString(), crn));
     }
+
+    private Long getTimestampThatThreeMonthsFromNow() {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java:203]: Misleading naming as the returned value is 3 months _before_ now

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java [hunk 13]
```diff
@@ -158,16 +165,24 @@ public StructuredEventContainer getStructuredEventsForStackByCrn(String crn, Lon
         return getEventsForUserWithResourceId(stack.getType().getResourceType(), stack.getId());
     }
 
+    public void deleteEntriesByResourceIdsOlderThanThreeMonths(Long resourceIds) {
+        repository.deleteRecordsOlderThan(resourceIds, getTimestampThatThreeMonthsFromNow());
+    }
+
+    public void deleteEntriesForTenantThatIsOlderThanThreeMonths(String tenantName) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java:172]: Misleading naming: the `tenantName` is actually the `accountId`

File: core/src/test/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    private ClusterComponentConfigProvider mockClusterComponentConfigProvider;
+
+    @Mock
+    private ClusterService mockClusterService;
+
+    private FinalizationCleanUpService underTest;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpServiceTest.java:40]: @ExtendWith

File: core/src/test/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    @Mock
+    private LegacyStructuredEventDBService mockStructuredEventDBService;
+
+    @Mock
+    private ClusterComponentConfigProvider mockClusterComponentConfigProvider;
+
+    @Mock
+    private ClusterService mockClusterService;
+
+    private FinalizationCleanUpService underTest;
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpServiceTest.java:36]: @InjectMocks

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpService.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        executeCleanUp(() -> clusterComponentConfigProvider.cleanUpDetachedEntries(clusterService.findIdsByDeletedStatus()), ClusterComponentHistory.class);
+    }
+
+    public void cleanUpStructuredEventsForStack(Long stackId) {
+        LOGGER.debug("About to start fetch and clean up structured events.");
+        executeCleanUp(() -> structuredEventDBService.deleteEntriesByResourceIdsOlderThanThreeMonths(stackId), StructuredEventEntity.class,
+                "that belogs to the following resource ID:" + stackId);
+    }
+
+    public void cleanUpStructuredEventsForTenant(String tenantName) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpService.java:41]: `tenantName` is misleading here, it should be `accountId`

File: cluster-api/src/main/java/com/sequenceiq/cloudbreak/repository/ClusterComponentHistoryRepository.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.Collection;
+
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponentHistory;
+import com.sequenceiq.cloudbreak.workspace.repository.EntityType;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.lang.NonNull;
+
+@Transactional(TxType.REQUIRED)
+@EntityType(entityClass = ClusterComponentHistory.class)
+public interface ClusterComponentHistoryRepository extends CrudRepository<ClusterComponentHistory, Long> {
```
Comment (bergerdenes) [inline cluster-api/src/main/java/com/sequenceiq/cloudbreak/repository/ClusterComponentHistoryRepository.java:15]: I don't really get this class, neither the ClusterComponentHistory.
It is a brand new entity, what is going to fill it? If it did not exist before, why is it needed to clean it up?

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java (hunk 7 above)
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/db/LegacyStructuredEventDBService.java:95]: I guess it looks more reasonable :)

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**